# Neural Networks

In [1]:
%pip install pandas matplotlib seaborn scikit-learn tensorflow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.9/10.9 MB 15.9 MB/s  0:00:00m0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.7/8.7 MB 33.1 MB/s  0:00:006m0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 20.3 MB/s  0:00:00m0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 572.6/572.6 MB 12.2 MB/s  0:00:25m0:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.8/6.8 MB 15.2 MB/s  0:00:00m eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 36.9 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.0/5.0 MB 20.5 MB/s  0:00:006m0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.0/5.0 MB 14.3 MB/s  0:00:006m0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 10.1 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 4.2 MB/s  0:00:00m eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.5/24.5 MB 19.5 MB/s  0:00:01m0:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 27.1 MB/s  0:00:00m0:00:01


In [1]:
from sklearn.preprocessing import StandardScaler
import pandas as pd

data = pd.read_csv('/workspaces/group-project-bas-team/data/cleaned_data.csv')

train = data[data['order_year']<=2021]
test = data[data['order_year']>2021]

data_limited_train = train.drop(['Title', 'ASIN/ISBN (Product Code)'], axis=1)
data_limited_test = test.drop(['Title', 'ASIN/ISBN (Product Code)'], axis=1)

X_train = data_limited_train.drop('Category', axis=1)
y_train = data_limited_train['Category']
X_train_scaled = StandardScaler().fit_transform(X_train)

X_test = data_limited_test.drop('Category', axis=1)
y_test = data_limited_test['Category']
X_test_scaled = StandardScaler().fit_transform(X_test)

In [3]:
X_train_scaled.shape[1]

153

In [4]:
y_train.max()

np.int64(1624)

In [2]:
import tensorflow as tf

model = tf.keras.Sequential()

model.add(tf.keras.layers.InputLayer(input_shape=(153, )))

model.add(tf.keras.layers.Dense(300, activation="relu")) 

model.add(tf.keras.layers.Dense(100, activation="relu"))

model.add(tf.keras.layers.Dense(1625, activation="softmax")) 

I0000 00:00:1775864435.760366   77278 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
I0000 00:00:1775864437.499371   77278 cpu_feature_guard.cc:227] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0000 00:00:1775864446.543435   77278 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
/usr/local/python/3.12.1/lib/python3.12/site-packages/keras/src/layers/core/input_layer.py:27: UserWarning: Argument `input_shape` is deprecated. Use `shape` instead.
  warnings.warn(
E0000 00:00:1775864449.279230   77278 cuda_platform.cc:52] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


In [6]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 300)            │        46,200 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 100)            │        30,100 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1625)           │       164,125 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 240,425 (939.16 KB)

 Trainable params: 240,425 (939.16 KB)

 Non-trainable params: 0 (0.00 B)

In [7]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

In [8]:
history = model.fit(X_train_scaled, y_train, epochs=30)

Epoch 1/30


W0000 00:00:1775853123.645715    4611 cpu_allocator_impl.cc:82] Allocation of 69556860 exceeds 10% of free system memory.


3552/3552 ━━━━━━━━━━━━━━━━━━━━ 12s 3ms/step - accuracy: 0.0591 - loss: 6.4504
Epoch 2/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0687 - loss: 6.0568
Epoch 3/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0763 - loss: 5.9418
Epoch 4/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 21s 3ms/step - accuracy: 0.0824 - loss: 5.8447
Epoch 5/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 20s 3ms/step - accuracy: 0.0871 - loss: 5.7554
Epoch 6/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 22s 3ms/step - accuracy: 0.0908 - loss: 5.6743
Epoch 7/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0940 - loss: 5.6002
Epoch 8/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0977 - loss: 5.5343
Epoch 9/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.1004 - loss: 5.4750
Epoch 10/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 19s 3ms/step - accuracy: 0.1021 - loss: 5.4225
Epoch 11/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.1047 - loss: 5.3758
Epoch 12/30
3552/3552 ━━━━━━━━

In [9]:
test_loss, test_acc = model.evaluate(X_test_scaled, y_test, verbose=2)
print('\nTest accuracy:', test_acc)

W0000 00:00:1775853520.466544    4611 cpu_allocator_impl.cc:82] Allocation of 26543052 exceeds 10% of free system memory.


1356/1356 - 4s - 3ms/step - accuracy: 0.0687 - loss: 5.9993

Test accuracy: 0.06873255968093872


Our initial neural network has an accuracy of about 6.9%, which is an improvement from our random forest model.

Next, we will attempt to build a wide & deep neural network

In [10]:
tf.keras.backend.clear_session()
tf.random.set_seed(42)

normalization_layer = tf.keras.layers.Normalization()

# two Dense layers with 30 neurons each, using the ReLU activation function
hidden_layer1 = tf.keras.layers.Dense(30, activation="relu")
hidden_layer2 = tf.keras.layers.Dense(30, activation="relu")

concat_layer = tf.keras.layers.Concatenate()

output_layer = tf.keras.layers.Dense(1625)


input_ = tf.keras.layers.Input(shape=X_train_scaled.shape[1:])

normalized = normalization_layer(input_)

hidden1 = hidden_layer1(normalized)
hidden2 = hidden_layer2(hidden1)

concat = concat_layer([normalized, hidden2])

output = output_layer(concat)

model = tf.keras.Model(inputs=[input_], outputs=[output])
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 153)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ normalization       │ (None, 153)       │        307 │ input_layer[0][0] │
│ (Normalization)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (Dense)       │ (None, 30)        │      4,620 │ normalization[0]… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_1 (Dense)     │ (None, 30)        │        930 │ dense[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate         │ (None, 183)       │          0 │ normalization[0]… │
│ (Concatenate)       │                   │            │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_2 (Dense)     │ (None, 1625)      │    299,000 │ concatenate[0][0] │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 304,857 (1.16 MB)

 Trainable params: 304,550 (1.16 MB)

 Non-trainable params: 307 (1.20 KB)

In [11]:
optimizer = tf.keras.optimizers.Adam(learning_rate=1e-3)
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

normalization_layer.adapt(X_train_scaled)
history = model.fit(X_train_scaled, y_train, epochs=20)
test_loss, test_acc = model.evaluate(X_test_scaled, y_test, verbose=2)
print('\nTest accuracy:', test_acc)

Epoch 1/20


W0000 00:00:1775853525.709696    4611 cpu_allocator_impl.cc:82] Allocation of 69556860 exceeds 10% of free system memory.


3552/3552 ━━━━━━━━━━━━━━━━━━━━ 15s 4ms/step - accuracy: 0.0018 - loss: 10.8478
Epoch 2/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.0018 - loss: 9.7746
Epoch 3/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.0018 - loss: 9.4219
Epoch 4/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.0018 - loss: 9.2080
Epoch 5/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 14s 4ms/step - accuracy: 0.0019 - loss: 9.1588
Epoch 6/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.0019 - loss: 9.1283
Epoch 7/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 21s 4ms/step - accuracy: 0.0020 - loss: 9.1006
Epoch 8/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.0019 - loss: 9.0295
Epoch 9/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.0020 - loss: 8.9518
Epoch 10/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 0.0019 - loss: 8.8985
Epoch 11/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 22s 4ms/step - accuracy: 0.0018 - loss: 8.8783
Epoch 12/20
3552/3552 ━━━━━━━

W0000 00:00:1775853804.737149    4611 cpu_allocator_impl.cc:82] Allocation of 26543052 exceeds 10% of free system memory.


1356/1356 - 4s - 3ms/step - accuracy: 0.0018 - loss: 8.9536

Test accuracy: 0.0017984367441385984


Very poor accuracy and high loss - need to revise

In [12]:
X_train.columns

Index(['Purchase Price Per Unit', 'Quantity', 'age', 'hispanic', 'education',
       'income', 'howmany', 'hh-size', 'how-oft', 'order_month',
       ...
       'life-changes_Lost a job ,Divorce',
       'life-changes_Lost a job ,Divorce,Moved place of residence',
       'life-changes_Lost a job ,Had a child',
       'life-changes_Lost a job ,Moved place of residence',
       'life-changes_Lost a job ,Moved place of residence,Became pregnant',
       'life-changes_Lost a job ,Moved place of residence,Became pregnant,Had a child',
       'life-changes_Lost a job ,Moved place of residence,Had a child',
       'life-changes_Moved place of residence',
       'life-changes_Moved place of residence,Became pregnant,Had a child',
       'life-changes_Moved place of residence,Had a child'],
      dtype='str', length=153)

In [13]:
corr_matrix = X_train.corr().abs()
top_corr = corr_matrix.unstack().sort_values(ascending=False)
top_corr = top_corr[top_corr < 1].drop_duplicates()
print(top_corr.head(20))

diabetes_No                                 diabetes_Yes                 0.999812
wheelchair_Yes                              wheelchair_No                0.998180
state_Alabama                               Shipping Address State_AL    0.989636
Shipping Address State_ME                   state_Maine                  0.987357
state_Iowa                                  Shipping Address State_IA    0.982405
sexual-orientation_heterosexual (straight)  sexual-orientation_LGBTQ+    0.978118
Shipping Address State_NE                   state_Nebraska               0.973867
Shipping Address State_MI                   state_Michigan               0.972118
state_Idaho                                 Shipping Address State_ID    0.971907
marijuana_Yes                               marijuana_No                 0.969490
Shipping Address State_MO                   state_Missouri               0.969445
Shipping Address State_KY                   state_Kentucky               0.966610
state_North Caro

In [14]:
# It appears that states and shipping address states are highly correlated - we will remove all shipping address state variables
# Same with the diabetes_No, wheelchair_No, Marijuana_No, sexual-orientation_heterosexual (straight), gender_female, and alcohol_No variables, which are all very highly correlated with other variables
# this will significantly limit the number of variables to subset, making it easier to build our wide and deep neural network

X_train = X_train.drop(X_train.filter(regex='^Shipping Address').columns, axis=1)
X_test = X_test.drop(X_test.filter(regex='^Shipping Address').columns, axis=1)

In [15]:
X_train = X_train.drop(columns = ['diabetes_No', 'wheelchair_No', 'marijuana_No', 'sexual-orientation_heterosexual (straight)', 
                                  'gender_Female', 'alcohol_No'])
X_test = X_test.drop(columns = ['diabetes_No', 'wheelchair_No', 'marijuana_No', 'sexual-orientation_heterosexual (straight)', 
                                  'gender_Female', 'alcohol_No'])

In [16]:
X_train.columns

Index(['Purchase Price Per Unit', 'Quantity', 'age', 'hispanic', 'education',
       'income', 'howmany', 'hh-size', 'how-oft', 'order_month', 'order_day',
       'order_year', 'cigarettes_I stopped in the recent past',
       'cigarettes_No', 'cigarettes_Yes',
       'alcohol_I stopped in the recent past', 'alcohol_Yes',
       'marijuana_I stopped in the recent past', 'marijuana_Yes',
       'diabetes_Yes', 'wheelchair_Yes', 'gender_Male', 'gender_Other',
       'sexual-orientation_LGBTQ+', 'state_Alabama', 'state_Arizona',
       'state_Arkansas', 'state_California', 'state_Colorado',
       'state_Connecticut', 'state_Delaware', 'state_Florida', 'state_Georgia',
       'state_Hawaii', 'state_Idaho', 'state_Illinois', 'state_Indiana',
       'state_Iowa', 'state_Kansas', 'state_Kentucky', 'state_Louisiana',
       'state_Maine', 'state_Maryland', 'state_Massachusetts',
       'state_Michigan', 'state_Minnesota', 'state_Missouri', 'state_Nebraska',
       'state_Nevada', 'state_New H

In [17]:
X_test.columns

Index(['Purchase Price Per Unit', 'Quantity', 'age', 'hispanic', 'education',
       'income', 'howmany', 'hh-size', 'how-oft', 'order_month', 'order_day',
       'order_year', 'cigarettes_I stopped in the recent past',
       'cigarettes_No', 'cigarettes_Yes',
       'alcohol_I stopped in the recent past', 'alcohol_Yes',
       'marijuana_I stopped in the recent past', 'marijuana_Yes',
       'diabetes_Yes', 'wheelchair_Yes', 'gender_Male', 'gender_Other',
       'sexual-orientation_LGBTQ+', 'state_Alabama', 'state_Arizona',
       'state_Arkansas', 'state_California', 'state_Colorado',
       'state_Connecticut', 'state_Delaware', 'state_Florida', 'state_Georgia',
       'state_Hawaii', 'state_Idaho', 'state_Illinois', 'state_Indiana',
       'state_Iowa', 'state_Kansas', 'state_Kentucky', 'state_Louisiana',
       'state_Maine', 'state_Maryland', 'state_Massachusetts',
       'state_Michigan', 'state_Minnesota', 'state_Missouri', 'state_Nebraska',
       'state_Nevada', 'state_New H

In [18]:
X_train

,Purchase Price Per Unit,Quantity,age,hispanic,education,income,howmany,hh-size,how-oft,order_month,...,"life-changes_Lost a job ,Divorce","life-changes_Lost a job ,Divorce,Moved place of residence","life-changes_Lost a job ,Had a child","life-changes_Lost a job ,Moved place of residence","life-changes_Lost a job ,Moved place of residence,Became pregnant","life-changes_Lost a job ,Moved place of residence,Became pregnant,Had a child","life-changes_Lost a job ,Moved place of residence,Had a child",life-changes_Moved place of residence,"life-changes_Moved place of residence,Became pregnant,Had a child","life-changes_Moved place of residence,Had a child"
0,7.98,1,3,1,3,2,1,1,1,12,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,13.99,1,3,1,3,2,1,1,1,12,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,10.45,1,3,1,3,2,1,1,1,12,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,10.00,1,3,1,3,2,1,1,1,12,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,10.99,1,3,1,3,2,1,1,1,2,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
156886,1089.99,1,2,0,2,2,1,2,1,12,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
156887,15.95,1,2,0,2,2,1,2,1,12,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
156888,8.99,1,2,0,2,2,1,2,1,12,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
156889,52.49,1,2,0,2,2,1,2,1,12,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0


In [19]:
X_train_scaled = StandardScaler().fit_transform(X_train)
X_test_scaled = StandardScaler().fit_transform(X_test)

In [20]:
# Rerunning our neural networks to see if removing highly correlated variables improved our accuracy

tf.keras.backend.clear_session()
tf.random.set_seed(42)

model = tf.keras.Sequential()

model.add(tf.keras.layers.InputLayer(input_shape=(98, )))

model.add(tf.keras.layers.Dense(300, activation="relu")) 

model.add(tf.keras.layers.Dense(100, activation="relu"))

model.add(tf.keras.layers.Dense(1625, activation="softmax")) 

/usr/local/python/3.12.1/lib/python3.12/site-packages/keras/src/layers/core/input_layer.py:27: UserWarning: Argument `input_shape` is deprecated. Use `shape` instead.
  warnings.warn(


In [21]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 300)            │        29,700 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 100)            │        30,100 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1625)           │       164,125 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 223,925 (874.71 KB)

 Trainable params: 223,925 (874.71 KB)

 Non-trainable params: 0 (0.00 B)

In [22]:
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

In [23]:
history = model.fit(X_train_scaled, y_train, epochs=30)

Epoch 1/30


W0000 00:00:1775853814.940294    4611 cpu_allocator_impl.cc:82] Allocation of 44552760 exceeds 10% of free system memory.


3552/3552 ━━━━━━━━━━━━━━━━━━━━ 12s 3ms/step - accuracy: 0.0571 - loss: 6.4680
Epoch 2/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0676 - loss: 6.0826
Epoch 3/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0743 - loss: 5.9842
Epoch 4/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0799 - loss: 5.8931
Epoch 5/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0850 - loss: 5.8068
Epoch 6/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0886 - loss: 5.7269
Epoch 7/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0923 - loss: 5.6540
Epoch 8/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0952 - loss: 5.5883
Epoch 9/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0979 - loss: 5.5298
Epoch 10/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.0998 - loss: 5.4782
Epoch 11/30
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 0.1016 - loss: 5.4328
Epoch 12/30
3552/3552 ━━━━━━━━

In [24]:
test_loss, test_acc = model.evaluate(X_test_scaled, y_test, verbose=2)
print('\nTest accuracy:', test_acc)

1356/1356 - 3s - 2ms/step - accuracy: 0.0695 - loss: 5.9546

Test accuracy: 0.06953955441713333


In [25]:
tf.keras.backend.clear_session()
tf.random.set_seed(42)

normalization_layer = tf.keras.layers.Normalization()

# two Dense layers with 30 neurons each, using the ReLU activation function
hidden_layer1 = tf.keras.layers.Dense(30, activation="relu")
hidden_layer2 = tf.keras.layers.Dense(30, activation="relu")

concat_layer = tf.keras.layers.Concatenate()

output_layer = tf.keras.layers.Dense(1625)


input_ = tf.keras.layers.Input(shape=X_train_scaled.shape[1:])

normalized = normalization_layer(input_)

hidden1 = hidden_layer1(normalized)
hidden2 = hidden_layer2(hidden1)

concat = concat_layer([normalized, hidden2])

output = output_layer(concat)

# Finally create the model, specifying inputs and outputs
model = tf.keras.Model(inputs=[input_], outputs=[output])
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 98)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ normalization       │ (None, 98)        │        197 │ input_layer[0][0] │
│ (Normalization)     │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense (Dense)       │ (None, 30)        │      2,970 │ normalization[0]… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_1 (Dense)     │ (None, 30)        │        930 │ dense[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ concatenate         │ (None, 128)       │          0 │ normalization[0]… │
│ (Concatenate)       │                   │            │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_2 (Dense)     │ (None, 1625)      │    209,625 │ concatenate[0][0] │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 213,722 (834.86 KB)

 Trainable params: 213,525 (834.08 KB)

 Non-trainable params: 197 (792.00 B)

In [26]:
optimizer = tf.keras.optimizers.Adam(learning_rate=1e-3)
model.compile(loss="sparse_categorical_crossentropy",
              optimizer="sgd",
              metrics=["accuracy"])

normalization_layer.adapt(X_train_scaled)
history = model.fit(X_train_scaled, y_train, epochs=20)
test_loss, test_acc = model.evaluate(X_test_scaled, y_test, verbose=2)
print('\nTest accuracy:', test_acc)

Epoch 1/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 12s 3ms/step - accuracy: 0.0010 - loss: 8.8313
Epoch 2/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 3.5194e-05 - loss: 8.0538
Epoch 3/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 3.5194e-05 - loss: 7.8997
Epoch 4/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 6.1590e-05 - loss: 7.8198
Epoch 5/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 8.7986e-05 - loss: 7.7054
Epoch 6/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 1.6717e-04 - loss: 7.8203
Epoch 7/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 1.8477e-04 - loss: 7.8924
Epoch 8/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 11s 3ms/step - accuracy: 2.1996e-04 - loss: 7.8497
Epoch 9/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 13s 4ms/step - accuracy: 2.1996e-04 - loss: 7.7933
Epoch 10/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 16s 4ms/step - accuracy: 2.2876e-04 - loss: 7.7458
Epoch 11/20
3552/3552 ━━━━━━━━━━━━━━━━━━━━ 19s 4ms/step - accuracy: 2.463

Next, we will try subsetting features so that the wide and deep components are trained on different features

In [3]:
# Wide: transactional features with likely direct/linear effect on category
wide_cols = [
    'Purchase Price Per Unit', 'Quantity',
    'order_month', 'order_day', 'order_year'
]

# Deep: demographic + behavioral features where interactions matter
deep_cols = [c for c in X_train.columns if c not in wide_cols]

print(f"Wide features : {len(wide_cols)}")
print(f"Deep features : {len(deep_cols)}")

Wide features : 5
Deep features : 148


In [4]:
wide_scaler = StandardScaler()
deep_scaler  = StandardScaler()

X_train_wide = wide_scaler.fit_transform(X_train[wide_cols])
X_test_wide  = wide_scaler.transform(X_test[wide_cols])

X_train_deep = deep_scaler.fit_transform(X_train[deep_cols])
X_test_deep  = deep_scaler.transform(X_test[deep_cols])

In [70]:
tf.keras.backend.clear_session()
tf.random.set_seed(42)

n_wide  = X_train_wide.shape[1]   # 5
n_deep  = X_train_deep.shape[1]   # 93
n_classes = 1625

input_wide = tf.keras.layers.Input(shape=(n_wide,),  name="wide_input")
input_deep = tf.keras.layers.Input(shape=(n_deep,),  name="deep_input")

# A single dense layer lets the model learn direct feature→class weights
wide_out = tf.keras.layers.Dense(64, activation="relu", name="wide_dense")(input_wide)

deep = tf.keras.layers.Dense(256, activation="relu", name="deep1")(input_deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep2")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(64, activation="relu", name="deep3")(deep)
deep_out = tf.keras.layers.Dropout(0.2)(deep)

# Concatenates then classifies
concat = tf.keras.layers.Concatenate(name="wide_deep_concat")([wide_out, deep_out])
output = tf.keras.layers.Dense(n_classes, activation="softmax", name="output")(concat)

model = tf.keras.Model(inputs=[input_wide, input_deep], outputs=output)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ deep_input          │ (None, 93)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep1 (Dense)       │ (None, 256)       │     24,064 │ deep_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 256)       │      1,024 │ deep1[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (Dropout)   │ (None, 256)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep2 (Dense)       │ (None, 128)       │     32,896 │ dropout[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep2[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_input          │ (None, 5)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep3 (Dense)       │ (None, 64)        │      8,256 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_dense (Dense)  │ (None, 64)        │        384 │ wide_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 64)        │          0 │ deep3[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_deep_concat    │ (None, 128)       │          0 │ wide_dense[0][0], │
│ (Concatenate)       │                   │            │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ output (Dense)      │ (None, 1625)      │    209,625 │ wide_deep_concat… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 276,761 (1.06 MB)

 Trainable params: 275,993 (1.05 MB)

 Non-trainable params: 768 (3.00 KB)

In [71]:
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
    metrics=["accuracy"]
)

callbacks = [
    tf.keras.callbacks.EarlyStopping(
        monitor="val_loss", patience=5, restore_best_weights=True
    ),
    tf.keras.callbacks.ReduceLROnPlateau(
        monitor="val_loss", factor=0.5, patience=3, min_lr=1e-6, verbose=1
    )
]

In [73]:
history = model.fit([X_train_wide, X_train_deep], y_train, epochs=100, callbacks=callbacks, validation_split=0.3)

Epoch 1/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0867 - loss: 5.5334 - val_accuracy: 0.0854 - val_loss: 5.5173 - learning_rate: 0.0010
Epoch 2/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0905 - loss: 5.4003 - val_accuracy: 0.0789 - val_loss: 5.6501 - learning_rate: 0.0010
Epoch 3/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 20s 8ms/step - accuracy: 0.0948 - loss: 5.3025 - val_accuracy: 0.0732 - val_loss: 5.7700 - learning_rate: 0.0010
Epoch 4/100
2480/2487 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.0940 - loss: 5.2407
Epoch 4: ReduceLROnPlateau reducing learning rate to 0.0005000000237487257.
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 15s 6ms/step - accuracy: 0.0962 - loss: 5.2322 - val_accuracy: 0.0777 - val_loss: 5.8659 - learning_rate: 0.0010
Epoch 5/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 27s 11ms/step - accuracy: 0.0996 - loss: 5.1371 - val_accuracy: 0.0760 - val_loss: 5.9190 - learning_rate: 5.0000e-04
Epoch 6/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 17s 7ms/step - accu

In [74]:
# Test model accuracy
test_loss, test_acc = model.evaluate([X_test_wide, X_test_deep], y_test, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

1356/1356 - 5s - 3ms/step - accuracy: 0.0644 - loss: 6.0153

Test accuracy: 0.0644


In [53]:
# Same model as above, but applying early stopping

tf.keras.backend.clear_session()
tf.random.set_seed(42)

n_wide  = X_train_wide.shape[1]   # 5
n_deep  = X_train_deep.shape[1]   # 93
n_classes = 1625

input_wide = tf.keras.layers.Input(shape=(n_wide,),  name="wide_input")
input_deep = tf.keras.layers.Input(shape=(n_deep,),  name="deep_input")

# A single dense layer lets the model learn direct feature→class weights
wide_out = tf.keras.layers.Dense(64, activation="relu", name="wide_dense")(input_wide)

deep = tf.keras.layers.Dense(256, activation="relu", name="deep1")(input_deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep2")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(64, activation="relu", name="deep3")(deep)
deep_out = tf.keras.layers.Dropout(0.2)(deep)

# Concatenates then classifies
concat = tf.keras.layers.Concatenate(name="wide_deep_concat")([wide_out, deep_out])
output = tf.keras.layers.Dense(n_classes, activation="softmax", name="output")(concat)

model = tf.keras.Model(inputs=[input_wide, input_deep], outputs=output)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ deep_input          │ (None, 93)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep1 (Dense)       │ (None, 256)       │     24,064 │ deep_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 256)       │      1,024 │ deep1[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (Dropout)   │ (None, 256)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep2 (Dense)       │ (None, 128)       │     32,896 │ dropout[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep2[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_input          │ (None, 5)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep3 (Dense)       │ (None, 64)        │      8,256 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_dense (Dense)  │ (None, 64)        │        384 │ wide_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 64)        │          0 │ deep3[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_deep_concat    │ (None, 128)       │          0 │ wide_dense[0][0], │
│ (Concatenate)       │                   │            │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ output (Dense)      │ (None, 1625)      │    209,625 │ wide_deep_concat… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 276,761 (1.06 MB)

 Trainable params: 275,993 (1.05 MB)

 Non-trainable params: 768 (3.00 KB)

In [54]:
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
    metrics=["accuracy"]
)

early_stop = tf.keras.callbacks.EarlyStopping(monitor="val_accuracy", patience=5, restore_best_weights=True)

In [55]:
history = model.fit([X_train_wide, X_train_deep], y_train, epochs=400, callbacks=early_stop, validation_split=0.3)

Epoch 1/400
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 21s 8ms/step - accuracy: 0.0655 - loss: 6.0691 - val_accuracy: 0.0637 - val_loss: 6.0522
Epoch 2/400
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 19s 8ms/step - accuracy: 0.0820 - loss: 5.6773 - val_accuracy: 0.0629 - val_loss: 6.0655
Epoch 3/400
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 16s 7ms/step - accuracy: 0.0882 - loss: 5.4947 - val_accuracy: 0.0641 - val_loss: 6.1009
Epoch 4/400
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 17s 7ms/step - accuracy: 0.0923 - loss: 5.3715 - val_accuracy: 0.0624 - val_loss: 6.1837
Epoch 5/400
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0955 - loss: 5.2805 - val_accuracy: 0.0630 - val_loss: 6.2274
Epoch 6/400
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0969 - loss: 5.2115 - val_accuracy: 0.0649 - val_loss: 6.2901
Epoch 7/400
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 17s 7ms/step - accuracy: 0.0973 - loss: 5.1601 - val_accuracy: 0.0635 - val_loss: 6.3600
Epoch 8/400
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0999 -

In [56]:
# Test model accuracy
test_loss, test_acc = model.evaluate([X_test_wide, X_test_deep], y_test, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

1356/1356 - 3s - 3ms/step - accuracy: 0.0579 - loss: 6.9355

Test accuracy: 0.0579


In [60]:
# Trying more layers and more neurons per layer

tf.keras.backend.clear_session()
tf.random.set_seed(42)

n_wide  = X_train_wide.shape[1]   # 5
n_deep  = X_train_deep.shape[1]   # 93
n_classes = 1625

input_wide = tf.keras.layers.Input(shape=(n_wide,),  name="wide_input")
input_deep = tf.keras.layers.Input(shape=(n_deep,),  name="deep_input")

# A single dense layer lets the model learn direct feature→class weights
wide_out = tf.keras.layers.Dense(256, activation="relu", name="wide_dense")(input_wide)

deep = tf.keras.layers.Dense(256, activation="relu", name="deep1")(input_deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep2")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep3")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep4")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep5")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep6")(deep)
deep_out = tf.keras.layers.Dropout(0.2)(deep)

# Concatenates then classifies
concat = tf.keras.layers.Concatenate(name="wide_deep_concat")([wide_out, deep_out])
output = tf.keras.layers.Dense(n_classes, activation="softmax", name="output")(concat)

model = tf.keras.Model(inputs=[input_wide, input_deep], outputs=output)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ deep_input          │ (None, 93)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep1 (Dense)       │ (None, 256)       │     24,064 │ deep_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 256)       │      1,024 │ deep1[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (Dropout)   │ (None, 256)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep2 (Dense)       │ (None, 128)       │     32,896 │ dropout[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep2[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep3 (Dense)       │ (None, 128)       │     16,512 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep3[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep4 (Dense)       │ (None, 128)       │     16,512 │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep4[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_3 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep5 (Dense)       │ (None, 128)       │     16,512 │ dropout_3[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep5[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_4 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_input          │ (None, 5)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep6 (Dense)       │ (None, 128)       │     16,512 │ dropout_4[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_dense (Dense)  │ (None, 256)       │      1,536 │ wide_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_5 (Dropout) │ (None, 128)       │          0 │ deep6[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_deep_concat    │ (None, 384)       │          0 │ wide_dense[0][0], │
│ (Concatenate)       │                   │            │ dropout_5[0][0] 

 Total params: 753,241 (2.87 MB)

 Trainable params: 751,705 (2.87 MB)

 Non-trainable params: 1,536 (6.00 KB)

In [61]:
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
    metrics=["accuracy"]
)

early_stop=tf.keras.callbacks.EarlyStopping(monitor="val_accuracy", patience=5, restore_best_weights=True)

In [64]:
history = model.fit([X_train_wide, X_train_deep], y_train, epochs=100, callbacks=early_stop, validation_split=0.3)


Epoch 1/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 27s 11ms/step - accuracy: 0.0644 - loss: 5.9789 - val_accuracy: 0.0646 - val_loss: 5.9600
Epoch 2/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 40s 10ms/step - accuracy: 0.0732 - loss: 5.7586 - val_accuracy: 0.0662 - val_loss: 5.9728
Epoch 3/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 28s 11ms/step - accuracy: 0.0782 - loss: 5.6247 - val_accuracy: 0.0626 - val_loss: 6.0478
Epoch 4/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 27s 11ms/step - accuracy: 0.0819 - loss: 5.5233 - val_accuracy: 0.0644 - val_loss: 6.1216
Epoch 5/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 27s 11ms/step - accuracy: 0.0849 - loss: 5.4376 - val_accuracy: 0.0621 - val_loss: 6.1827
Epoch 6/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 27s 11ms/step - accuracy: 0.0867 - loss: 5.3730 - val_accuracy: 0.0602 - val_loss: 6.2915
Epoch 7/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 28s 11ms/step - accuracy: 0.0884 - loss: 5.3196 - val_accuracy: 0.0615 - val_loss: 6.3611


In [65]:
test_loss, test_acc = model.evaluate([X_test_wide, X_test_deep], y_test, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

1356/1356 - 7s - 5ms/step - accuracy: 0.0546 - loss: 6.1474

Test accuracy: 0.0546


In [66]:
# Same model, but increasing number of epochs and using early stopping

tf.keras.backend.clear_session()
tf.random.set_seed(42)

n_wide  = X_train_wide.shape[1]   # 5
n_deep  = X_train_deep.shape[1]   # 93
n_classes = 1625

input_wide = tf.keras.layers.Input(shape=(n_wide,),  name="wide_input")
input_deep = tf.keras.layers.Input(shape=(n_deep,),  name="deep_input")

# A single dense layer lets the model learn direct feature→class weights
wide_out = tf.keras.layers.Dense(256, activation="relu", name="wide_dense")(input_wide)

deep = tf.keras.layers.Dense(256, activation="relu", name="deep1")(input_deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep2")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep3")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep4")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep5")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="relu", name="deep6")(deep)
deep_out = tf.keras.layers.Dropout(0.2)(deep)

# Concatenates then classifies
concat = tf.keras.layers.Concatenate(name="wide_deep_concat")([wide_out, deep_out])
output = tf.keras.layers.Dense(n_classes, activation="softmax", name="output")(concat)

model = tf.keras.Model(inputs=[input_wide, input_deep], outputs=output)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ deep_input          │ (None, 93)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep1 (Dense)       │ (None, 256)       │     24,064 │ deep_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 256)       │      1,024 │ deep1[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (Dropout)   │ (None, 256)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep2 (Dense)       │ (None, 128)       │     32,896 │ dropout[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep2[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep3 (Dense)       │ (None, 128)       │     16,512 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep3[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep4 (Dense)       │ (None, 128)       │     16,512 │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep4[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_3 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep5 (Dense)       │ (None, 128)       │     16,512 │ dropout_3[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep5[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_4 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_input          │ (None, 5)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep6 (Dense)       │ (None, 128)       │     16,512 │ dropout_4[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_dense (Dense)  │ (None, 256)       │      1,536 │ wide_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_5 (Dropout) │ (None, 128)       │          0 │ deep6[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_deep_concat    │ (None, 384)       │          0 │ wide_dense[0][0], │
│ (Concatenate)       │                   │            │ dropout_5[0][0] 

 Total params: 753,241 (2.87 MB)

 Trainable params: 751,705 (2.87 MB)

 Non-trainable params: 1,536 (6.00 KB)

In [67]:
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
    metrics=["accuracy"]
)

early_stop = tf.keras.callbacks.EarlyStopping(monitor="val_accuracy", patience=5, restore_best_weights=True)

In [68]:
history = model.fit([X_train_wide, X_train_deep], y_train, epochs=400, callbacks=early_stop, validation_split=0.2)

Epoch 1/400
2842/2842 ━━━━━━━━━━━━━━━━━━━━ 34s 10ms/step - accuracy: 0.0618 - loss: 6.0928 - val_accuracy: 0.0663 - val_loss: 5.9262
Epoch 2/400
2842/2842 ━━━━━━━━━━━━━━━━━━━━ 30s 10ms/step - accuracy: 0.0715 - loss: 5.7953 - val_accuracy: 0.0690 - val_loss: 5.9145
Epoch 3/400
2842/2842 ━━━━━━━━━━━━━━━━━━━━ 39s 10ms/step - accuracy: 0.0769 - loss: 5.6627 - val_accuracy: 0.0627 - val_loss: 5.9965
Epoch 4/400
2842/2842 ━━━━━━━━━━━━━━━━━━━━ 41s 10ms/step - accuracy: 0.0814 - loss: 5.5572 - val_accuracy: 0.0643 - val_loss: 6.0452
Epoch 5/400
2842/2842 ━━━━━━━━━━━━━━━━━━━━ 30s 10ms/step - accuracy: 0.0832 - loss: 5.4769 - val_accuracy: 0.0622 - val_loss: 6.1265
Epoch 6/400
2842/2842 ━━━━━━━━━━━━━━━━━━━━ 29s 10ms/step - accuracy: 0.0849 - loss: 5.4087 - val_accuracy: 0.0622 - val_loss: 6.1682
Epoch 7/400
2842/2842 ━━━━━━━━━━━━━━━━━━━━ 28s 10ms/step - accuracy: 0.0874 - loss: 5.3538 - val_accuracy: 0.0635 - val_loss: 6.2516


In [69]:
test_loss, test_acc = model.evaluate([X_test_wide, X_test_deep], y_test, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

1356/1356 - 6s - 4ms/step - accuracy: 0.0568 - loss: 6.0805

Test accuracy: 0.0568


Our best model based on test accuracy still appears to be our original wide & deep network, with a test accuracy of 6.44%. We will revert back to this model and attempt to change the activation function.

In [75]:
tf.keras.backend.clear_session()
tf.random.set_seed(42)

n_wide  = X_train_wide.shape[1]   # 5
n_deep  = X_train_deep.shape[1]   # 93
n_classes = 1625

input_wide = tf.keras.layers.Input(shape=(n_wide,),  name="wide_input")
input_deep = tf.keras.layers.Input(shape=(n_deep,),  name="deep_input")

# A single dense layer lets the model learn direct feature→class weights
wide_out = tf.keras.layers.Dense(64, activation="elu", name="wide_dense")(input_wide)

deep = tf.keras.layers.Dense(256, activation="elu", name="deep1")(input_deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="elu", name="deep2")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(64, activation="elu", name="deep3")(deep)
deep_out = tf.keras.layers.Dropout(0.2)(deep)

# Concatenates then classifies
concat = tf.keras.layers.Concatenate(name="wide_deep_concat")([wide_out, deep_out])
output = tf.keras.layers.Dense(n_classes, activation="softmax", name="output")(concat)

model = tf.keras.Model(inputs=[input_wide, input_deep], outputs=output)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ deep_input          │ (None, 93)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep1 (Dense)       │ (None, 256)       │     24,064 │ deep_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 256)       │      1,024 │ deep1[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (Dropout)   │ (None, 256)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep2 (Dense)       │ (None, 128)       │     32,896 │ dropout[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep2[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_input          │ (None, 5)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep3 (Dense)       │ (None, 64)        │      8,256 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_dense (Dense)  │ (None, 64)        │        384 │ wide_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 64)        │          0 │ deep3[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_deep_concat    │ (None, 128)       │          0 │ wide_dense[0][0], │
│ (Concatenate)       │                   │            │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ output (Dense)      │ (None, 1625)      │    209,625 │ wide_deep_concat… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 276,761 (1.06 MB)

 Trainable params: 275,993 (1.05 MB)

 Non-trainable params: 768 (3.00 KB)

In [76]:
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
    metrics=["accuracy"]
)

callbacks = [
    tf.keras.callbacks.EarlyStopping(
        monitor="val_loss", patience=5, restore_best_weights=True
    ),
    tf.keras.callbacks.ReduceLROnPlateau(
        monitor="val_loss", factor=0.5, patience=3, min_lr=1e-6, verbose=1
    )
]

In [77]:
history = model.fit([X_train_wide, X_train_deep], y_train, epochs=100, callbacks = callbacks, validation_split=0.3)

Epoch 1/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 30s 11ms/step - accuracy: 0.0714 - loss: 6.0287 - val_accuracy: 0.0557 - val_loss: 6.0871 - learning_rate: 0.0010
Epoch 2/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 25s 10ms/step - accuracy: 0.0837 - loss: 5.6384 - val_accuracy: 0.0549 - val_loss: 6.0789 - learning_rate: 0.0010
Epoch 3/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 22s 9ms/step - accuracy: 0.0877 - loss: 5.5111 - val_accuracy: 0.0577 - val_loss: 6.1179 - learning_rate: 0.0010
Epoch 4/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 25s 10ms/step - accuracy: 0.0904 - loss: 5.4400 - val_accuracy: 0.0570 - val_loss: 6.1364 - learning_rate: 0.0010
Epoch 5/100
2481/2487 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.0916 - loss: 5.3978
Epoch 5: ReduceLROnPlateau reducing learning rate to 0.0005000000237487257.
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 25s 10ms/step - accuracy: 0.0925 - loss: 5.3861 - val_accuracy: 0.0585 - val_loss: 6.1588 - learning_rate: 0.0010
Epoch 6/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 19s 8ms/step - accur

In [78]:
test_loss, test_acc = model.evaluate([X_test_wide, X_test_deep], y_test, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

1356/1356 - 5s - 4ms/step - accuracy: 0.0535 - loss: 6.0386

Test accuracy: 0.0535


In [79]:
tf.keras.backend.clear_session()
tf.random.set_seed(42)

n_wide  = X_train_wide.shape[1]   # 5
n_deep  = X_train_deep.shape[1]   # 93
n_classes = 1625

input_wide = tf.keras.layers.Input(shape=(n_wide,),  name="wide_input")
input_deep = tf.keras.layers.Input(shape=(n_deep,),  name="deep_input")

# A single dense layer lets the model learn direct feature→class weights
wide_out = tf.keras.layers.Dense(64, activation="selu", name="wide_dense")(input_wide)

deep = tf.keras.layers.Dense(256, activation="selu", name="deep1")(input_deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="selu", name="deep2")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(64, activation="selu", name="deep3")(deep)
deep_out = tf.keras.layers.Dropout(0.2)(deep)

# Concatenates then classifies
concat = tf.keras.layers.Concatenate(name="wide_deep_concat")([wide_out, deep_out])
output = tf.keras.layers.Dense(n_classes, activation="softmax", name="output")(concat)

model = tf.keras.Model(inputs=[input_wide, input_deep], outputs=output)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ deep_input          │ (None, 93)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep1 (Dense)       │ (None, 256)       │     24,064 │ deep_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 256)       │      1,024 │ deep1[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (Dropout)   │ (None, 256)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep2 (Dense)       │ (None, 128)       │     32,896 │ dropout[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep2[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_input          │ (None, 5)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep3 (Dense)       │ (None, 64)        │      8,256 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_dense (Dense)  │ (None, 64)        │        384 │ wide_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 64)        │          0 │ deep3[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_deep_concat    │ (None, 128)       │          0 │ wide_dense[0][0], │
│ (Concatenate)       │                   │            │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ output (Dense)      │ (None, 1625)      │    209,625 │ wide_deep_concat… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 276,761 (1.06 MB)

 Trainable params: 275,993 (1.05 MB)

 Non-trainable params: 768 (3.00 KB)

In [80]:
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
    metrics=["accuracy"]
)

callbacks = [
    tf.keras.callbacks.EarlyStopping(
        monitor="val_loss", patience=5, restore_best_weights=True
    ),
    tf.keras.callbacks.ReduceLROnPlateau(
        monitor="val_loss", factor=0.5, patience=3, min_lr=1e-6, verbose=1
    )
]

In [81]:
history = model.fit([X_train_wide, X_train_deep], y_train, epochs=100, callbacks = callbacks, validation_split=0.3)

Epoch 1/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 24s 9ms/step - accuracy: 0.0700 - loss: 6.0182 - val_accuracy: 0.0556 - val_loss: 6.1188 - learning_rate: 0.0010
Epoch 2/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 17s 7ms/step - accuracy: 0.0821 - loss: 5.6446 - val_accuracy: 0.0516 - val_loss: 6.1270 - learning_rate: 0.0010
Epoch 3/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0865 - loss: 5.5305 - val_accuracy: 0.0547 - val_loss: 6.1330 - learning_rate: 0.0010
Epoch 4/100
2484/2487 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.0862 - loss: 5.4809
Epoch 4: ReduceLROnPlateau reducing learning rate to 0.0005000000237487257.
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0885 - loss: 5.4643 - val_accuracy: 0.0554 - val_loss: 6.1450 - learning_rate: 0.0010
Epoch 5/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 19s 7ms/step - accuracy: 0.0926 - loss: 5.3462 - val_accuracy: 0.0545 - val_loss: 6.1700 - learning_rate: 5.0000e-04
Epoch 6/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 20s 8ms/step - accur

In [ ]:
test_loss, test_acc = model.evaluate([X_test_wide, X_test_deep], y_test, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

1356/1356 - 3s - 2ms/step - accuracy: 0.0514 - loss: 6.0815

Test accuracy: 0.0514


In [5]:
tf.keras.backend.clear_session()
tf.random.set_seed(42)

n_wide  = X_train_wide.shape[1]   # 5
n_deep  = X_train_deep.shape[1]   # 93
n_classes = 1625

input_wide = tf.keras.layers.Input(shape=(n_wide,),  name="wide_input")
input_deep = tf.keras.layers.Input(shape=(n_deep,),  name="deep_input")

# A single dense layer lets the model learn direct feature→class weights
wide_out = tf.keras.layers.Dense(64, activation="gelu", name="wide_dense")(input_wide)

deep = tf.keras.layers.Dense(256, activation="gelu", name="deep1")(input_deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="gelu", name="deep2")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(64, activation="gelu", name="deep3")(deep)
deep_out = tf.keras.layers.Dropout(0.2)(deep)

# Concatenates then classifies
concat = tf.keras.layers.Concatenate(name="wide_deep_concat")([wide_out, deep_out])
output = tf.keras.layers.Dense(n_classes, activation="softmax", name="output")(concat)

model = tf.keras.Model(inputs=[input_wide, input_deep], outputs=output)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ deep_input          │ (None, 148)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep1 (Dense)       │ (None, 256)       │     38,144 │ deep_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 256)       │      1,024 │ deep1[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (Dropout)   │ (None, 256)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep2 (Dense)       │ (None, 128)       │     32,896 │ dropout[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep2[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_input          │ (None, 5)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep3 (Dense)       │ (None, 64)        │      8,256 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_dense (Dense)  │ (None, 64)        │        384 │ wide_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 64)        │          0 │ deep3[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_deep_concat    │ (None, 128)       │          0 │ wide_dense[0][0], │
│ (Concatenate)       │                   │            │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ output (Dense)      │ (None, 1625)      │    209,625 │ wide_deep_concat… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 290,841 (1.11 MB)

 Trainable params: 290,073 (1.11 MB)

 Non-trainable params: 768 (3.00 KB)

In [6]:
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
    metrics=["accuracy"]
)

callbacks = [
    tf.keras.callbacks.EarlyStopping(
        monitor="val_loss", patience=5, restore_best_weights=True
    ),
    tf.keras.callbacks.ReduceLROnPlateau(
        monitor="val_loss", factor=0.5, patience=3, min_lr=1e-6, verbose=1
    )
]

In [7]:
history = model.fit([X_train_wide, X_train_deep], y_train, epochs=100, callbacks = callbacks, validation_split=0.3)

Epoch 1/100


W0000 00:00:1775864489.096532   77278 cpu_allocator_impl.cc:82] Allocation of 47098336 exceeds 10% of free system memory.


2483/2487 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.0580 - loss: 6.2961

W0000 00:00:1775864508.239398   77278 cpu_allocator_impl.cc:82] Allocation of 20185424 exceeds 10% of free system memory.


2487/2487 ━━━━━━━━━━━━━━━━━━━━ 22s 8ms/step - accuracy: 0.0693 - loss: 6.0578 - val_accuracy: 0.0630 - val_loss: 6.0937 - learning_rate: 0.0010
Epoch 2/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0862 - loss: 5.6427 - val_accuracy: 0.0590 - val_loss: 6.1079 - learning_rate: 0.0010
Epoch 3/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 23s 8ms/step - accuracy: 0.0923 - loss: 5.4689 - val_accuracy: 0.0598 - val_loss: 6.1216 - learning_rate: 0.0010
Epoch 4/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.0937 - loss: 5.3734
Epoch 4: ReduceLROnPlateau reducing learning rate to 0.0005000000237487257.
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0957 - loss: 5.3625 - val_accuracy: 0.0606 - val_loss: 6.1570 - learning_rate: 0.0010
Epoch 5/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 21s 8ms/step - accuracy: 0.1002 - loss: 5.2361 - val_accuracy: 0.0590 - val_loss: 6.1792 - learning_rate: 5.0000e-04
Epoch 6/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 17s 7ms/step - accuracy: 0.1024 

In [8]:
test_loss, test_acc = model.evaluate([X_test_wide, X_test_deep], y_test, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

W0000 00:00:1775864859.848553   77278 cpu_allocator_impl.cc:82] Allocation of 25675632 exceeds 10% of free system memory.


1356/1356 - 4s - 3ms/step - accuracy: 0.0568 - loss: 6.0898

Test accuracy: 0.0568


In [9]:
tf.keras.backend.clear_session()
tf.random.set_seed(42)

n_wide  = X_train_wide.shape[1]   # 5
n_deep  = X_train_deep.shape[1]   # 93
n_classes = 1625

input_wide = tf.keras.layers.Input(shape=(n_wide,),  name="wide_input")
input_deep = tf.keras.layers.Input(shape=(n_deep,),  name="deep_input")

# A single dense layer lets the model learn direct feature→class weights
wide_out = tf.keras.layers.Dense(64, activation="swish", name="wide_dense")(input_wide)

deep = tf.keras.layers.Dense(256, activation="swish", name="deep1")(input_deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="swish", name="deep2")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(64, activation="swish", name="deep3")(deep)
deep_out = tf.keras.layers.Dropout(0.2)(deep)

# Concatenates then classifies
concat = tf.keras.layers.Concatenate(name="wide_deep_concat")([wide_out, deep_out])
output = tf.keras.layers.Dense(n_classes, activation="softmax", name="output")(concat)

model = tf.keras.Model(inputs=[input_wide, input_deep], outputs=output)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ deep_input          │ (None, 148)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep1 (Dense)       │ (None, 256)       │     38,144 │ deep_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 256)       │      1,024 │ deep1[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (Dropout)   │ (None, 256)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep2 (Dense)       │ (None, 128)       │     32,896 │ dropout[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep2[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_input          │ (None, 5)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep3 (Dense)       │ (None, 64)        │      8,256 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_dense (Dense)  │ (None, 64)        │        384 │ wide_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 64)        │          0 │ deep3[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_deep_concat    │ (None, 128)       │          0 │ wide_dense[0][0], │
│ (Concatenate)       │                   │            │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ output (Dense)      │ (None, 1625)      │    209,625 │ wide_deep_concat… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 290,841 (1.11 MB)

 Trainable params: 290,073 (1.11 MB)

 Non-trainable params: 768 (3.00 KB)

In [10]:
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
    metrics=["accuracy"]
)

callbacks = [
    tf.keras.callbacks.EarlyStopping(
        monitor="val_loss", patience=5, restore_best_weights=True
    ),
    tf.keras.callbacks.ReduceLROnPlateau(
        monitor="val_loss", factor=0.5, patience=3, min_lr=1e-6, verbose=1
    )
]

In [11]:
history = model.fit([X_train_wide, X_train_deep], y_train, epochs=100, callbacks = callbacks, validation_split=0.3)

Epoch 1/100


W0000 00:00:1775864872.701167   77278 cpu_allocator_impl.cc:82] Allocation of 47098336 exceeds 10% of free system memory.


2487/2487 ━━━━━━━━━━━━━━━━━━━━ 23s 9ms/step - accuracy: 0.0692 - loss: 6.0504 - val_accuracy: 0.0649 - val_loss: 6.0738 - learning_rate: 0.0010
Epoch 2/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 24s 10ms/step - accuracy: 0.0843 - loss: 5.6412 - val_accuracy: 0.0619 - val_loss: 6.1020 - learning_rate: 0.0010
Epoch 3/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 23s 9ms/step - accuracy: 0.0911 - loss: 5.4799 - val_accuracy: 0.0588 - val_loss: 6.1252 - learning_rate: 0.0010
Epoch 4/100
2483/2487 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.0910 - loss: 5.3956
Epoch 4: ReduceLROnPlateau reducing learning rate to 0.0005000000237487257.
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 20s 8ms/step - accuracy: 0.0934 - loss: 5.3818 - val_accuracy: 0.0581 - val_loss: 6.1429 - learning_rate: 0.0010
Epoch 5/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 16s 6ms/step - accuracy: 0.0998 - loss: 5.2594 - val_accuracy: 0.0593 - val_loss: 6.1559 - learning_rate: 5.0000e-04
Epoch 6/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.1013

In [12]:
test_loss, test_acc = model.evaluate([X_test_wide, X_test_deep], y_test, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

W0000 00:00:1775865021.836908   77278 cpu_allocator_impl.cc:82] Allocation of 25675632 exceeds 10% of free system memory.


1356/1356 - 5s - 3ms/step - accuracy: 0.0532 - loss: 6.0897

Test accuracy: 0.0532


In [13]:
tf.keras.backend.clear_session()
tf.random.set_seed(42)

n_wide  = X_train_wide.shape[1]   # 5
n_deep  = X_train_deep.shape[1]   # 93
n_classes = 1625

input_wide = tf.keras.layers.Input(shape=(n_wide,),  name="wide_input")
input_deep = tf.keras.layers.Input(shape=(n_deep,),  name="deep_input")

# A single dense layer lets the model learn direct feature→class weights
wide_out = tf.keras.layers.Dense(64, activation="mish", name="wide_dense")(input_wide)

deep = tf.keras.layers.Dense(256, activation="mish", name="deep1")(input_deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(128, activation="mish", name="deep2")(deep)
deep = tf.keras.layers.BatchNormalization()(deep)
deep = tf.keras.layers.Dropout(0.3)(deep)

deep = tf.keras.layers.Dense(64, activation="mish", name="deep3")(deep)
deep_out = tf.keras.layers.Dropout(0.2)(deep)

# Concatenates then classifies
concat = tf.keras.layers.Concatenate(name="wide_deep_concat")([wide_out, deep_out])
output = tf.keras.layers.Dense(n_classes, activation="softmax", name="output")(concat)

model = tf.keras.Model(inputs=[input_wide, input_deep], outputs=output)
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ deep_input          │ (None, 148)       │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep1 (Dense)       │ (None, 256)       │     38,144 │ deep_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalization │ (None, 256)       │      1,024 │ deep1[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout (Dropout)   │ (None, 256)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep2 (Dense)       │ (None, 128)       │     32,896 │ dropout[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ batch_normalizatio… │ (None, 128)       │        512 │ deep2[0][0]       │
│ (BatchNormalizatio… │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_1 (Dropout) │ (None, 128)       │          0 │ batch_normalizat… │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_input          │ (None, 5)         │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ deep3 (Dense)       │ (None, 64)        │      8,256 │ dropout_1[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_dense (Dense)  │ (None, 64)        │        384 │ wide_input[0][0]  │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dropout_2 (Dropout) │ (None, 64)        │          0 │ deep3[0][0]       │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ wide_deep_concat    │ (None, 128)       │          0 │ wide_dense[0][0], │
│ (Concatenate)       │                   │            │ dropout_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ output (Dense)      │ (None, 1625)      │    209,625 │ wide_deep_concat… │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 290,841 (1.11 MB)

 Trainable params: 290,073 (1.11 MB)

 Non-trainable params: 768 (3.00 KB)

In [14]:
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3),
    metrics=["accuracy"]
)

callbacks = [
    tf.keras.callbacks.EarlyStopping(
        monitor="val_loss", patience=5, restore_best_weights=True
    ),
    tf.keras.callbacks.ReduceLROnPlateau(
        monitor="val_loss", factor=0.5, patience=3, min_lr=1e-6, verbose=1
    )
]

In [15]:
history = model.fit([X_train_wide, X_train_deep], y_train, epochs=100, callbacks = callbacks, validation_split=0.3)

Epoch 1/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 21s 8ms/step - accuracy: 0.0703 - loss: 6.0394 - val_accuracy: 0.0625 - val_loss: 6.0747 - learning_rate: 0.0010
Epoch 2/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accuracy: 0.0866 - loss: 5.6241 - val_accuracy: 0.0581 - val_loss: 6.0959 - learning_rate: 0.0010
Epoch 3/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 19s 6ms/step - accuracy: 0.0923 - loss: 5.4712 - val_accuracy: 0.0564 - val_loss: 6.1135 - learning_rate: 0.0010
Epoch 4/100
2482/2487 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.0936 - loss: 5.3846
Epoch 4: ReduceLROnPlateau reducing learning rate to 0.0005000000237487257.
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 17s 7ms/step - accuracy: 0.0950 - loss: 5.3724 - val_accuracy: 0.0558 - val_loss: 6.1251 - learning_rate: 0.0010
Epoch 5/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 19s 7ms/step - accuracy: 0.0999 - loss: 5.2487 - val_accuracy: 0.0560 - val_loss: 6.1363 - learning_rate: 5.0000e-04
Epoch 6/100
2487/2487 ━━━━━━━━━━━━━━━━━━━━ 18s 7ms/step - accur

In [16]:
test_loss, test_acc = model.evaluate([X_test_wide, X_test_deep], y_test, verbose=2)
print(f"\nTest accuracy: {test_acc:.4f}")

1356/1356 - 3s - 2ms/step - accuracy: 0.0580 - loss: 6.0730

Test accuracy: 0.0580
